lib/tbot/config/migrate.go

```diff
@@ -0,0 +1,418 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"time"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/exp/slices"
+	"gopkg.in/yaml.v3"
+
+	"github.com/gravitational/teleport/lib/tbot/bot"
+)
+
+type destinationMixinV1 struct {
+	Directory *DestinationDirectory `yaml:"directory"`
+	Memory    *DestinationMemory    `yaml:"memory"`
+}
+
+func (c *destinationMixinV1) migrate() (bot.Destination, error) {
+	switch {
+	case c.Memory != nil && c.Directory != nil:
+		return nil, trace.BadParameter("both 'memory' and 'directory' cannot be specified")
+	case c.Memory != nil:
+		return c.Memory, nil
+	case c.Directory != nil:
+		return c.Directory, nil
+	default:
+		return nil, trace.BadParameter("at least one of `memory' and 'directory' must be specified")
+	}
+}
+
+type storageConfigV1 struct {
+	Mixin destinationMixinV1 `yaml:",inline"`
+}
+
+func (c *storageConfigV1) migrate() (*StorageConfig, error) {
+	dest, err := c.Mixin.migrate()
+	if err != nil {
+		return nil, trace.Wrap(err, "migrating destination mixin")
+	}
+	if dest == nil {
```

Comment:
Is this case allowed? 

`(c *destinationMixinV1) migrate()` returns either non-nil dst or an error

Reply:
This was a hold-over that I missed in refactor - good catch. bca696c984 removes this.